### PR TITLE
Fix method ambiguities in SparseArrays

### PR DESCRIPTION
**Pull request by tkf**
I added a few methods to resolve method ambiguity in SparseArrays.jl.  I also included the test using `detect_ambiguities` as I did for LinearAlgebra.jl in #28749 and #28816.

### CODE REVIEW

File: stdlib/SparseArrays/src/higherorderfns.jl [hunk 1]
```diff
@@ -1024,9 +1021,13 @@ end
...
 @inline capturescalars(f, mixedargs::Tuple{Ref{Type{T}}, Vararg{Any}}) where {T} =
     capturescalars((args...)->f(T, args...), Base.tail(mixedargs))
+@inline capturescalars(f, mixedargs::Tuple{Ref{Type{T}}, Ref{Type{S}}, Vararg{Any}}) where {T, S} =
+    # This definition is identical to the one above and necessary only for
+    # avoiding method ambiguity.
+    capturescalars((args...)->f(T, args...), Base.tail(mixedargs))
 @inline capturescalars(f, mixedargs::Tuple{SparseVecOrMat, Ref{Type{T}}, Vararg{Any}}) where {T} =
     capturescalars((a1, args...)->f(a1, T, args...), (mixedargs[1], Base.tail(Base.tail(mixedargs))...))
-@inline capturescalars(f, mixedargs::Tuple{Union{Ref,AbstractArray{0}}, Ref{Type{T}}, Vararg{Any}}) where {T} =
+@inline capturescalars(f, mixedargs::Tuple{Union{Ref,AbstractArray{<:Any,0}}, Ref{Type{T}}, Vararg{Any}}) where {T} =
```
Comment (tkf) [inline stdlib/SparseArrays/src/higherorderfns.jl]: This was irrelevant to this PR but I suppose `AbstractArray{0}` was a typo?

File: stdlib/SparseArrays/test/higherorderfns.jl [hunk 2]
```diff
@@ -632,4 +632,11 @@ end
     @test minimum(sparse([1, 2], [1, 2], ones(Int32, 2)), dims = 1) isa Matrix
 end
 
+@testset "Issue #30118" begin
+    @test ((_, x) -> x).(Int, spzeros(3)) == spzeros(3)
+    @test ((_, _, x) -> x).(Int, Int, spzeros(3)) == spzeros(3)
+    @test ((_, _, _, x) -> x).(Int, Int, Int, spzeros(3)) == spzeros(3)
+    @test_broken ((_, _, _, _, x) -> x).(Int, Int, Int, Int, spzeros(3)) == spzeros(3)
```
Comment (tkf) [inline stdlib/SparseArrays/test/higherorderfns.jl]: It looks like `capturescalars` has type inference issue when it has to recurse a lot (> 3).  I wounder if `capturescalars` could be implemented more naturally (for humans and maybe for Julia as well?) if you used `@generated` function.  Any reason why `capturescalars` is implemented based on recursion?

File: stdlib/SparseArrays/src/higherorderfns.jl [hunk 3]
```diff
@@ -1006,10 +1007,6 @@ end
     _copyto!(parevalf, dest, passedsrcargstup...)
 end
 
-struct CapturedScalars{F, Args, Order}
```
Comment (tkf) [inline stdlib/SparseArrays/src/higherorderfns.jl]: `git grep CapturedScalars` shows me that there is no code using it.  Maybe some remnant from experimenting with `capturescalars`?  This was irrelevant to this PR but I thought it's OK to remove it here (as I'm touching `capturescalars`).

File: stdlib/SparseArrays/src/higherorderfns.jl [hunk 4]
```diff
@@ -973,6 +973,7 @@ function _copy(f, args...)
     parevalf, passedargstup = capturescalars(f, args)
     return _copy(parevalf, passedargstup...)
 end
+_copy(f) = throw(MethodError(_copy, (f)))  # avoid method ambiguity
```
Comment (timholy) [inline stdlib/SparseArrays/src/higherorderfns.jl]: `(f) === f`, should be `(f,)`

**Review by timholy — COMMENTED** [review]
Thanks for doing this.

**Review by timholy — APPROVED** [review]

**Pull request merged**